console/services/market_app/component.py

```diff
@@ -0,0 +1,220 @@
+        self.ports = list(ports)
+        self.volumes = list(volumes)
+        self.config_files = list(config_files)
+        self.probe = probe
+        self.extend_info = extend_info
+        self.monitors = list(monitors)
+        self.graphs = list(graphs)
+        self.component_deps = []
+        self.volume_deps = []
+        self.app_config_groups = []
+        self.service_group_rel = service_group_rel
+
+    def set_changes(self, changes, governance_mode):
+        """
+        Set changes to the component
+        """
+        update_funcs = self._create_update_funcs()
+        for key in update_funcs:
+            if not changes.get(key):
+                continue
+            update_func = update_funcs[key]
+            update_func(changes.get(key))
+        self._ensure_port_envs(governance_mode)
+
+    def _ensure_port_envs(self, governance_mode):
+        # filter out the old port envs
+        envs = [env for env in self.envs if env.container_port != 0]
+        # create outer envs for every port
+        for port in self.ports:
+            envs.extend(self._create_envs_4_ports(port, governance_mode))
+        self.envs = envs
+
+    def _create_envs_4_ports(self, port: TenantServicesPort, governance_mode):
+        port_alias = self.component.service_alias.upper()
+        host_value = "127.0.0.1" if governance_mode == GovernanceModeEnum.BUILD_IN_SERVICE_MESH.name else port.k8s_service_name
+        host_env = self._create_port_env(port, "连接地址", port_alias + "_HOST", host_value)
+        port_env = self._create_port_env(port, "端口", port_alias + "_PORT", str(port.container_port))
+        return [host_env, port_env]
+
+    def _create_port_env(self, port: TenantServicesPort, name, attr_name, attr_value):
+        return TenantServiceEnvVar(
+            tenant_id=self.component.tenant_id,
+            service_id=self.component.component_id,
+            container_port=port.container_port,
+            name=name,
+            attr_name=attr_name,
+            attr_value=attr_value,
+            is_change=False,
+            scope="outer",
+            create_time=datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
+        )
+
+    def _create_update_funcs(self):
+        return {
+            "deploy_version": self._update_deploy_version,
+            "app_version": self._update_version,
+            # "image": self.dummy_func,  # TODO(huangrh)
+            "envs": self._update_inner_envs,
+            "connect_infos": self._update_outer_envs,
+            "ports": self._update_ports,
+            "volumes": self._update_volumes,
+            "probe": self._update_probe,
+            # "plugins": self._update_plugins,
+            "component_graphs": self._update_component_graphs,
+            "component_monitors": self._update_component_monitors,
+        }
+
+    def _update_deploy_version(self, dv):
+        if not dv["is_change"]:
+            return
+        self.component.deploy_version = dv["new"]
```

Comment:
The `dv` may not be a reliable data structure. It requires strict judgment. The other update function is similar.